flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/StagewiseSchedulingStrategy.java

```diff
@@ -0,0 +1,156 @@
+
+    @Override
+    public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+        maybeScheduleVerticesOneByOne(
+                SchedulingStrategyUtils.getVerticesFromIds(schedulingTopology, verticesToRestart));
+    }
+
+    @Override
+    public void onExecutionStateChange(
+            ExecutionVertexID executionVertexId, ExecutionState executionState) {
+        if (executionState == ExecutionState.FINISHED) {
+            SchedulingExecutionVertex executionVertex =
+                    schedulingTopology.getVertex(executionVertexId);
+
+            Set<SchedulingExecutionVertex> consumerVertices =
+                    IterableUtils.toStream(executionVertex.getProducedResults())
+                            .flatMap(partition -> IterableUtils.toStream(partition.getConsumers()))
+                            .collect(Collectors.toSet());
+
+            maybeScheduleVerticesOneByOne(consumerVertices);
+        }
+    }
+
+    @Override
+    public void onPartitionConsumable(IntermediateResultPartitionID resultPartitionId) {}
+
+    @Override
+    public void notifySchedulingTopologyUpdated(
+            SchedulingTopology schedulingTopology, List<ExecutionVertexID> newlyAddedVertices) {
+        checkState(schedulingTopology == this.schedulingTopology);
+        pendingVertices.addAll(
+                SchedulingStrategyUtils.getVerticesFromIds(
+                        schedulingTopology, new HashSet<>(newlyAddedVertices)));
+    }
+
+    private void maybeScheduleVerticesOneByOne(final Set<SchedulingExecutionVertex> vertices) {
+
+        final Set<SchedulingExecutionVertex> allCandidates = new HashSet<>(vertices);
+        allCandidates.addAll(
+                pendingVertices.stream()
+                        .filter(vertex -> vertex.getState() == ExecutionState.CREATED)
+                        .collect(Collectors.toSet()));
+
+        final Set<ExecutionVertexID> verticesToDeploy =
+                allCandidates.stream()
+                        .filter(this::areVertexInputsAllConsumable)
+                        .map(
+                                vertex -> {
+                                    checkState(vertex.getState() == ExecutionState.CREATED);
+                                    return vertex.getId();
+                                })
+                        .collect(Collectors.toSet());
+
+        if (!verticesToDeploy.isEmpty()) {
+            final List<ExecutionVertexDeploymentOption> vertexDeploymentOptions =
+                    SchedulingStrategyUtils
+                            .createExecutionVertexDeploymentOptionsInTopologicalOrder(
+                                    schedulingTopology, verticesToDeploy, id -> deploymentOption);
+
+            vertexDeploymentOptions.forEach(
+                    option ->
+                            schedulerOperations.allocateSlotsAndDeploy(
+                                    Collections.singletonList(option)));
+
+            pendingVertices.clear();
```

Comment:
I think the `clear()` should be invoked once `pendingVertices` are added into `allCandidates`.

Reply:
I will fix it.